src/python/pants/backend/docgen/tasks/markdown_to_html_utils.py

```diff
@@ -66,57 +72,91 @@ def choose_include_text(s, params, source_path):
   :param params: string like "start-at=foo&end-at=bar"
   :param source_path: path to source .md. Useful in error messages
   """
-  lines = s.splitlines()
-  start_after = None
-  start_at = None
-  end_before = None
-  end_at = None
+  source_lines = s.splitlines()
+  params_dict = {}
 
   for term in params.split("&"):
     if '=' in term:
       param, value = [p.strip() for p in term.split('=', 1)]
     else:
       param, value = term.strip(), ''
     if not param: continue
-    if param == "start-after":
-      start_after = value
-    elif param == "start-at":
-      start_at = value
-    elif param == "end-before":
-      end_before = value
-    elif param == "end-at":
-      end_at = value
+    if param not in INCLUDE_RECOGNIZED_PARAMS:
+      raise TaskError('Invalid include directive "{0}" in params "{1}"'
+                      ' from {2}'.format(param, params, source_path))
+    params_dict[param] = value
+
+  chosen_lines = []
+  try:
+    if 'lines' in params_dict:
+      lines_arg = params_dict['lines']
+      chosen_lines = include_by_line_range(source_lines, lines_arg, params_dict)
     else:
-      raise TaskError('Invalid include directive "{0}"'
-                      ' in {1}'.format(params, source_path))
+      chosen_lines = include_by_text_match(source_lines, params_dict)
+  except TaskError as include_error:
+    raise TaskError('Error: {0} (in params "{1}" from {2})'
+                    .format(include_error, params, source_path))
+  return '\n'.join(chosen_lines)
 
+
+def include_by_line_range(source_lines, line_range_arg, params_dict):
+  # 'lines' should be the only parameter specified
+  conflicts = [(p, params_dict[p]) for p in params_dict if p != 'lines']
+  if len(conflicts) > 0:
+    error_msgs = ', '.join(['"{0}={1}"'.format(p, v) for (p, v) in conflicts])
+    raise TaskError('"lines" directive conflicts with other directives: [{0}]'
+                    .format(error_msgs))
+
+  # validate the line range
+  lines_match = LINES_PATTERN.match(line_range_arg)
+  if lines_match is None:
+    raise TaskError('Invalid line range "{0}" (should match "{1}") for "lines"'
+                    .format(line_range_arg, LINES_PATTERN.pattern))
+  [line_start, line_end] = (int(g) for g in lines_match.groups())
+  # LINES_PATTERN does not allow 0 or negative numbers
+  if line_end < line_start:
+    raise TaskError('Invalid end line "{0}" (should be >= {1}, the start line)'
```

Comment:
All of these errors should mention which source file is to blame for the error.

Reply:
EDIT: see my response to your inline comment above on pre-parsing `params_dict` which will make all of this moot. Response kept to explain motivation.

I agree. I tried to avoid long repetitive format strings in `include_by_*` with the try/except block at the bottom of `choose_include_text`. It currently catches and re-raises any `TaskError`, and adds the `params` used in the `!inc` call and the `source_path` (the file with the error) to the error string. If this logic isn't obvious, we could make another exception class to throw in `include_by_*`, or make a small helper function to format `!inc` exception messages, or just not worry about long format strings.

In #4982 I mentioned wanting to eventually either make `start-at`/etc start throwing if it doesn't find the search text, or to add differently-named versions of those parameters which are more strict. If we want to throw more exceptions or add/change `!inc` parameters eventually it seems to make sense to have `choose_include_text` handle adding context such as `params` and `source_path`. But I have no issue with longer format strings if re-raising exceptions leads to unclear control flow.